### PR TITLE
Feature GitHub apiversion header

### PR DESCRIPTION
**Pull request by Atharva-Shinde**
Fixes: #28161

any inputs?
cc: @petr-muller @jihoon-seo

### CODE REVIEW

**Comment by stevekuznetsov** [comment]
Upgrading an API version is going to be done by the maintainers of this project, as breaking changes between versions will require changes to the code handling the API calls. It seems like the currently-supported API version should be a constant in the Prow codebase and we should pass it without validation to GitHub - checking to see if the version is still valid will simply be done by GitHub when serving our call, no need to do it manually.

File: prow/github/client.go [hunk 1]
```diff
@@ -1257,6 +1261,23 @@ func (c *client) authHeader() string {
 	return fmt.Sprintf("Bearer %s", token)
 }
 
+func (c *client) apiVersion() string {
+	resp, err := http.Get("https://api.github.com/versions")
```
Comment (petr-muller) [inline prow/github/client.go:1265]: We totally do not want to do such `GET` on all calls by the client... The version validity check is a neat idea but it should be done just once (at the time of client creation, I guess), through the client call chain itself (so that we get various logging, retry logic etc) and with proper error handling.

I think it would be totally acceptable for this PR to simply hardcode `2022-11-28` (the only valid version at this time) into a const and send a header with that. That would declare "prow client is using specific API version" which IMO totally fine. I'd probably also added a log message to the client creation that would log the version used.

You could invest time into making the API version configurable by e.g. a constructor taking a specific api version (then you could also include a method for verifying that the version is supported), but I think that at time we'd also want to actually think how different API versions would be supported in the client (=how to actually deal with differences?). At this point in time, that's not needed, and we also do not know whether the new API versions will even be relevant to Prow (we do not know how GitHub actually plans to break their APIs :) ), so I'm inclining towards calling [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) and vote for just harcoding the version now.

File: prow/github/client.go [hunk 2]
```diff
@@ -1257,6 +1261,23 @@ func (c *client) authHeader() string {
...
+	if err != nil {
+		fmt.Print(err)
```
Comment (petr-muller) [inline prow/github/client.go:1267]: Commenting just for the sake of completeness, see my verbose comment above otherwise about why I think we should not even have this check.

1. Errors should be always handled, not just written out. Sometimes the handling is "stop what you are doing and return the error so that something above in the chain handles it"
2. You basically never want to continue when something returns `err` - you must not rely the other returned value  (`resp` in this case).

**Comment by Atharva-Shinde** [comment]
thanks on the reviews, will work on it

**Comment by Atharva-Shinde** [comment]
Any thoughts on this?

https://docs.github.com/en/rest/overview/api-versions?apiVersion=2022-11-28#specifying-an-api-version
> Requests without the X-GitHub-Api-Version header will default to use the 2022-11-28 version.

**Comment by stevekuznetsov** [comment]
The reason you'd want to explicitly set the API version is that it's part of your contract with the server - either the server honors your version, and your client code works correctly, or the server explicitly tells you that it cannot honor your request. Not setting a version whatsoever and letting it default to something takes away that control from you - what happens when the default version that's used when no header is set changes? If your client is only capable of handling the previous version, your code will fail in some relatively incomprehensible way - failing to decode a JSON body, setting the wrong query parameters, etc.

**Comment by Atharva-Shinde** [comment]
Understood thanks!

**Comment by petr-muller** [comment]
Also GitHub [asked nicely](https://github.blog/2022-11-28-to-infinity-and-beyond-enabling-the-future-of-githubs-rest-api-with-api-versioning/) :D :

> We encourage integrators to update their integration to send the new X-GitHub-Api-Version: 2022-11-28 header.

**Comment by Atharva-Shinde** [comment]
hello again, 
can I get a little help here with the unit test :)

tldr: So apparently the fakeHttpClient is formatting the header string request hence causing the unit tests to fail.


Here's a case I tried while debugging:
in client.go:  `req.Header.Set("X-GITHUB-VERsion", githubApiVersion)` 

in client_test.go: 
expectedHeader for Empty token, no auth header: `"X-GITHUB-VERsion": []string{"2022-11-28"}`
expectedHeader for Token, auth header:  `"Authorization": []string{"Bearer sup"}, "X-GITHUB-VERsion": []string{"2022-11-28"}`

here's the log

```
Running tool: /usr/local/go/bin/go test -timeout 30s -run ^TestAuthHeaderGetsSet$ k8s.io/test-infra/prow/github

=== RUN   TestAuthHeaderGetsSet
=== PAUSE TestAuthHeaderGetsSet
=== CONT  TestAuthHeaderGetsSet
=== RUN   TestAuthHeaderGetsSet/Empty_token,_no_auth_header
time="2023-01-06T00:24:47+05:30" level=info msg="Using GitHub REST API Version: 2022-11-28"
    /Users/atharva/Desktop/test-infra/prow/github/client_test.go:3016: expected header differs from actual:   http.Header{
                "Accept":           {"application/vnd.github.v3+json"},
        -       "X-GITHUB-VERsion": {"2022-11-28"},
        +       "X-Github-Version": {"2022-11-28"},
          }
=== RUN   TestAuthHeaderGetsSet/Token,_auth_header
time="2023-01-06T00:24:47+05:30" level=info msg="Using GitHub REST API Version: 2022-11-28"
    /Users/atharva/Desktop/test-infra/prow/github/client_test.go:3016: expected header differs from actual:   http.Header{
                "Accept":           {"application/vnd.github.v3+json"},
                "Authorization":    {"Bearer sup"},
        -       "X-GITHUB-VERsion": {"2022-11-28"},
        +       "X-Github-Version": {"2022-11-28"},
          }
--- FAIL: TestAuthHeaderGetsSet (0.00s)
    --- FAIL: TestAuthHeaderGetsSet/Empty_token,_no_auth_header (0.00s)
    --- FAIL: TestAuthHeaderGetsSet/Token,_auth_header (0.00s)
FAIL
FAIL    k8s.io/test-infra/prow/github   0.371s


> Test run finished at 06/01/2023, 00:24:47 <
```

File: prow/github/client.go [hunk 3]
```diff
@@ -1182,6 +1183,8 @@ func (c *client) doRequest(ctx context.Context, method, path, accept, org string
 	if err != nil {
 		return nil, fmt.Errorf("failed creating new request: %w", err)
 	}
+	req.Header.Set("X-GitHub-Api-Version", githubApiVersion)
```
Comment (MadhavJivrajani) [inline prow/github/client.go:1186]: Responding to https://github.com/kubernetes/test-infra/pull/28340#issuecomment-1372618144 here.

For whatever reason, GitHub chose to call the header `X-GitHub-Api-Version` and not `X-Github-Api-Version`. The reason that it is getting changed to the latter is because `Set()` internally canonicalizes the header name ([ref](https://pkg.go.dev/net/http#Header.Set)). For rules of canonicalization, see [here](https://pkg.go.dev/net/textproto#CanonicalMIMEHeaderKey).

According to the docs:
> To use non-canonical keys, assign to the map directly.

So, to make this work, you'll have to do something like:
```suggestion
	req.Header["X-GitHub-Api-Version"] = []string{githubApiVersion}
```

Comment (Atharva-Shinde) [reply prow/github/client.go:1186]: thanks @MadhavJivrajani for the inputs :)

Comment (Atharva-Shinde) [reply prow/github/client.go]: that's a TIL

**Review by petr-muller — APPROVED** [review]
Thanks for the work!
